File: pr007.py
#!/usr/bin/env python2
# -*- coding: utf-8 -*-
#

'''
Problem 7
By listing the first six prime numbers: 2, 3, 5, 7, 11, and 13,
we can see that the 6th prime is 13.

What is the 10,001st prime number?
'''

NUM = 10001

def has_factor(num, lst):
    for l in lst:
        if num % l == 0:
            return True
    return False

def primes(num):
    '''return list of primes has num elements'''
    prs = []
    n = 0
    val = 1  # we start at '2' (not include '1')
    while n < num:
        val += 1
        if has_factor(val, prs):
            continue
        prs.append(val)
        n += 1
    return prs

def main():
    print primes(NUM)[-1]

if __name__ == '__main__':
    main()

File: pr007.hs
{-
Problem 7
By listing the first six prime numbers: 2, 3, 5, 7, 11, and 13,
we can see that the 6th prime is 13.

What is the 10,001st prime number?
-}

module Main where

has_factor :: Int -> [Int] -> Bool
has_factor x ns = or $ map (\n -> x `mod` n == 0) ns

prime :: [Int]
prime = [x | x <- [2..], not (has_factor x [2..(x-1)])]

main :: IO ()
main = print (head (drop 10000 prime))